src/query/storage/m3/storage.go

```diff
@@ -240,65 +241,57 @@ func (s *m3storage) CompleteTags(
 	}
 
 	// TODO: instead of aggregating locally, have the DB aggregate it before
-	// sending results back
+	// sending results back.
 	fetchQuery := &storage.FetchQuery{
 		TagMatchers: query.TagMatchers,
+		// NB: complete tags matches every tag from the start of time until now
+		Start: time.Time{},
+		End:   time.Now(),
 	}
+
 	results, cleanup, err := s.SearchCompressed(ctx, fetchQuery, options)
+	defer cleanup()
 	if err != nil {
 		return nil, err
 	}
 
-	defer func() { _ = cleanup() }()
-
-	tagsToTagValues := make(map[string]map[string]struct{})
+	accumulatedTags := storage.NewCompleteTagsResultBuilder(query.CompleteNameOnly)
 	for _, elem := range results {
-		for elem.Iter.Next() {
-			tag := elem.Iter.Current()
-			tagName := tag.Name.String()
-
-			values, ok := tagsToTagValues[tagName]
-			if !ok {
-				values = make(map[string]struct{})
-				tagsToTagValues[tagName] = values
+		it := elem.Iter
+		tags := make([]storage.CompletedTag, 0, it.Len())
+		for i := 0; it.Next(); i++ {
+			tag := it.Current()
+			name := tag.Name.Bytes()
+			found := false
+			for _, filterName := range query.FilterNameTags {
```

Comment:
consider separating this into Any() function ? https://gobyexample.com/collection-functions